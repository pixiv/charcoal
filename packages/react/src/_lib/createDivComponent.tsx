import { DetailedHTMLProps, DO_NOT_USE_OR_YOU_WILL_BE_FIRED_CALLBACK_REF_RETURN_VALUES, forwardRef, ForwardRefExoticComponent, HTMLAttributes, RefAttributes, RefObject } from 'react'
import { useClassNames } from './useClassNames'

export function createDivComponent(mainClassName: string): ForwardRefExoticComponent<Omit<Omit<DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>, "ref"> & {
  ref?: ((instance: HTMLDivElement | null) => void | DO_NOT_USE_OR_YOU_WILL_BE_FIRED_CALLBACK_REF_RETURN_VALUES[keyof DO_NOT_USE_OR_YOU_WILL_BE_FIRED_CALLBACK_REF_RETURN_VALUES]) | RefObject<HTMLDivElement> | null | undefined
}, "ref"> & RefAttributes<HTMLDivElement>> {
  return forwardRef<HTMLDivElement, React.ComponentPropsWithRef<'div'>>(
    function DivComponent({ className, ...props }, ref) {
      const classNames = useClassNames(mainClassName, className)
      return <div className={classNames} ref={ref} {...props} />
    }
  )
}
