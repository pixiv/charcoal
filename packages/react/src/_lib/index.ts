/**
 * 今後ポートされる予定の汎用的な関数群
 */

/**
 * Function used to assert a given code path is unreachable
 */
export function unreachable(): never
/**
 * Function used to assert a given code path is unreachable.
 * Very useful for ensuring switches are exhaustive:
 *
 * ```ts
 * switch (a.type) {
 *   case Types.A:
 *   case Types.B:
 *     break
 *   default:
 *     unreachable(a) // will cause a build error if there was
 *                    // a Types.C that was not checked
 * }
 * ```
 *
 * @param value Value to be asserted as unreachable
 */
// NOTE: Uses separate overloads, _not_ `value?: never`, to not allow `undefined` to be passed

export function unreachable(value: never): never
export function unreachable(value?: never): never {
  throw new Error(
    arguments.length === 0
      ? 'unreachable'
      : `unreachable (${JSON.stringify(value)})`
  )
}

/**
 * 複数のrefをマージする。
 *
 * forwardRefで受け取ったrefと、コンポーネント内で定義したrefを同じ要素につけたいケースなどで使う
 */
export function mergeRefs<T>(...refs: React.Ref<T>[]): React.RefCallback<T> {
  return (value) => {
    for (const ref of refs) {
      if (typeof ref === 'function') {
        ref(value)
      } else if (ref !== null) {
        // TODO: use RefObject
        // https://react.dev/blog/2024/04/25/react-19-upgrade-guide#:~:text=%3D%201%3B-,MutableRef,-is%20now%20deprecated
        ;(ref as React.MutableRefObject<T | null>).current = value
      }
    }
  }
}

export function countCodePointsInString(string: string): number {
  // [...string] とするとproduction buildで動かなくなる
  // cf. https://twitter.com/f_subal/status/1497214727511891972
  return Array.from(string).length
}
