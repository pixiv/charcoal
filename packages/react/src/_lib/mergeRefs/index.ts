import React from 'react'

/**
 * Custom implementation of mergeRefs to replace @react-aria/utils
 * Merges multiple refs into a single ref callback
 */
export function mergeRefs<T>(
  ...refs: (React.Ref<T> | undefined)[]
): React.RefCallback<T> {
  return (value: T | null) => {
    refs.forEach((ref) => {
      if (ref == null) {
        return
      }

      if (typeof ref === 'function') {
        ref(value)
      } else if (typeof ref === 'object') {
        ;(ref as React.MutableRefObject<T | null>).current = value
      }
    })
  }
}
