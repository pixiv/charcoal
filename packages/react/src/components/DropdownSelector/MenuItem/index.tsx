import React from 'react'
import ListItem, { CustomJSXElement, ListItemProps } from '../ListItem'
import { useMenuItemHandleKeyDown } from './internals/useMenuItemHandleKeyDown'

export type MenuItemProps<T extends CustomJSXElement = never> = {
  value?: string
  disabled?: boolean
} & ListItemProps<T>

/**
 * 上下キーでフォーカス移動でき、エンターキーで選択できるリストの項目
 * 基本的に`<MenuList>`, `<MenuGroup>`と合わせて使用する
 */
export default function MenuItem<T extends CustomJSXElement>(
  props: MenuItemProps<T>
) {
  const { children, as, ...rest } = props
  const [handleKeyDown, setContextValue] = useMenuItemHandleKeyDown(props.value)
  return (
    <ListItem
      {...rest}
      as={as as CustomJSXElement}
      data-key={props.value}
      onKeyDown={handleKeyDown}
      onClick={props.disabled === true ? undefined : setContextValue}
      tabIndex={-1}
      aria-disabled={props.disabled}
    >
      {props.children}
    </ListItem>
  )
}
