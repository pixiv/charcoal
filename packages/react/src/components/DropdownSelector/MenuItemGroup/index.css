.charcoal-menu-item-group {
  display: block;
}

.charcoal-menu-item-group > span {
  display: block;
  color: var(--charcoal-color-text-tertiary-default);
  font-size: 12px;
  font-weight: bold;
  padding: 12px 0 8px 16px;
}

.charcoal-menu-item-group > ul {
  padding-left: 0;
  margin: 0;
  box-sizing: border-box;
  list-style: none;
  overflow: hidden;
}
