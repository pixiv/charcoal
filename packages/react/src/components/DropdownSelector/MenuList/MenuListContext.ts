import { RefObject, createContext } from 'react'

type MenuListContextType = {
  root?: RefObject<HTMLUListElement>
  value?: string
  values?: string[]
  setValue: (v: string) => void
}

export const MenuListContext = createContext<MenuListContextType>({
  root: undefined,
  value: '',
  values: [],
  setValue: (_v: string) => {
    // empty
  },
})
