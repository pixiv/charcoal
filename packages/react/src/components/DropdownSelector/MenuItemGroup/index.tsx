import React from 'react'
import styled from 'styled-components'
import MenuItem from '../MenuItem'
import { Divider } from '../Divider'

type MenuItemGroupChild = React.ReactElement<typeof MenuItem | typeof Divider>

export type MenuItemGroupProps = {
  text: string
  children: MenuItemGroupChild | MenuItemGroupChild[]
}

/**
 * 項目のリストを分類する見出しをつけるコンテナ要素
 */
export default function MenuItemGroup(props: MenuItemGroupProps) {
  return (
    <StyledLi role="presentation">
      <TextSpan>{props.text}</TextSpan>
      <StyledUl role="group">{props.children}</StyledUl>
    </StyledLi>
  )
}

const TextSpan = styled.span`
  display: block;
  color: ${({ theme }) => theme.color.text3};
  font-size: 12px;
  font-weight: bold;
  padding: 12px 0 8px 16px;
`

const StyledUl = styled.ul`
  padding-left: 0;
  margin: 0;
  box-sizing: border-box;
  list-style: none;
  overflow: hidden;
`

const StyledLi = styled.li`
  display: block;
`
