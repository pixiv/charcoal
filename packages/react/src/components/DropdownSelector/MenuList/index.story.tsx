import React from 'react'
import { action } from '@storybook/addon-actions'
import { Story } from '../../../_lib/compat'
import MenuList, { MenuListProps } from '.'
import MenuItem from '../MenuItem'
import MenuItemGroup from '../MenuItemGroup'

export default {
  title: 'DropdownSelector/MenuList',
  component: MenuList,
}

function makeList(n: number, offset = 0) {
  return [...(Array(n) as undefined[])].map((_, i) => {
    const v = i + offset
    return (
      <MenuItem key={v} value={v.toString()}>
        Menu {v}
      </MenuItem>
    )
  })
}

export const Basic: Story<MenuListProps> = () => {
  return (
    <>
      <MenuList onChange={action('onChange')}>{makeList(10)}</MenuList>
    </>
  )
}

export const Disabled: Story<MenuListProps> = () => {
  return (
    <>
      <MenuList onChange={action('onChange')}>
        <MenuItem value="1">MenuItem</MenuItem>
        <MenuItem value="2" disabled>
          Disabled MenuItem
        </MenuItem>
      </MenuList>
    </>
  )
}

export const Group: Story<MenuListProps> = () => {
  return (
    <MenuList onChange={action('onChange')} value="1">
      <MenuItemGroup text="Section1">{makeList(5)}</MenuItemGroup>
      <MenuItemGroup text="Section2">{makeList(5, 5)}</MenuItemGroup>
    </MenuList>
  )
}
