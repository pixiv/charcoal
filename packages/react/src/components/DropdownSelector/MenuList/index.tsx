import React, { useRef } from 'react'
import styled from 'styled-components'
import { MenuListContext } from './MenuListContext'
import { getValuesRecursive } from './internals/getValuesRecursive'
import MenuItem from '../MenuItem'
import { Divider } from '../Divider'
import MenuItemGroup from '../MenuItemGroup'

type MenuListChild = React.ReactElement<
  typeof MenuItem | typeof MenuItemGroup | typeof Divider
>

export type MenuListChildren = MenuListChild | MenuListChild[]

export type MenuListProps = {
  children: MenuListChildren
  value?: string
  onChange?: (v: string) => void
}

/**
 * 上下キーでフォーカス移動でき、エンターキーで選択できるリストの項目
 * 基本的に`<MenuItem>`, `<MenuGroup>`と合わせて使用する
 */
export default function MenuList(props: MenuListProps) {
  const root = useRef(null)
  const values: string[] = []
  getValuesRecursive(props.children, values)

  return (
    <StyledUl ref={root}>
      <MenuListContext.Provider
        value={{
          value: props.value ?? '',
          root,
          values,
          setValue: (v) => {
            props.onChange?.(v)
          },
        }}
      >
        {props.children}
      </MenuListContext.Provider>
    </StyledUl>
  )
}

const StyledUl = styled.ul`
  padding: 0;
  margin: 0;
`
