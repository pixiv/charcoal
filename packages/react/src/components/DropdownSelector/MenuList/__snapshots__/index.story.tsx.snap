// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots DropdownSelector/MenuList Basic 1`] = `
.c1 {
  list-style: none;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 40px;
  cursor: pointer;
  outline: none;
  padding-right: 16px;
  padding-left: 16px;
}

.c2:disabled,
.c2[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c2:hover,
.c2:focus,
.c2:focus-within {
  background-color: var(--charcoal-surface3);
}

.c0 {
  padding: 0;
  margin: 0;
}

<div
  data-dark={false}
>
  <ul
    className="c0"
  >
    <li
      className="c1"
      role="option"
    >
      <div
        className="c2"
        data-key="0"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
        value="0"
      >
        Menu 
        0
      </div>
    </li>
    <li
      className="c1"
      role="option"
    >
      <div
        className="c2"
        data-key="1"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
        value="1"
      >
        Menu 
        1
      </div>
    </li>
    <li
      className="c1"
      role="option"
    >
      <div
        className="c2"
        data-key="2"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
        value="2"
      >
        Menu 
        2
      </div>
    </li>
    <li
      className="c1"
      role="option"
    >
      <div
        className="c2"
        data-key="3"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
        value="3"
      >
        Menu 
        3
      </div>
    </li>
    <li
      className="c1"
      role="option"
    >
      <div
        className="c2"
        data-key="4"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
        value="4"
      >
        Menu 
        4
      </div>
    </li>
    <li
      className="c1"
      role="option"
    >
      <div
        className="c2"
        data-key="5"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
        value="5"
      >
        Menu 
        5
      </div>
    </li>
    <li
      className="c1"
      role="option"
    >
      <div
        className="c2"
        data-key="6"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
        value="6"
      >
        Menu 
        6
      </div>
    </li>
    <li
      className="c1"
      role="option"
    >
      <div
        className="c2"
        data-key="7"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
        value="7"
      >
        Menu 
        7
      </div>
    </li>
    <li
      className="c1"
      role="option"
    >
      <div
        className="c2"
        data-key="8"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
        value="8"
      >
        Menu 
        8
      </div>
    </li>
    <li
      className="c1"
      role="option"
    >
      <div
        className="c2"
        data-key="9"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
        value="9"
      >
        Menu 
        9
      </div>
    </li>
  </ul>
</div>
`;

exports[`Storyshots DropdownSelector/MenuList Disabled 1`] = `
.c1 {
  list-style: none;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 40px;
  cursor: pointer;
  outline: none;
  padding-right: 16px;
  padding-left: 16px;
}

.c2:disabled,
.c2[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c2:hover,
.c2:focus,
.c2:focus-within {
  background-color: var(--charcoal-surface3);
}

.c0 {
  padding: 0;
  margin: 0;
}

<div
  data-dark={false}
>
  <ul
    className="c0"
  >
    <li
      className="c1"
      role="option"
    >
      <div
        className="c2"
        data-key="1"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
        value="1"
      >
        MenuItem
      </div>
    </li>
    <li
      className="c1"
      role="option"
    >
      <div
        aria-disabled={true}
        className="c2"
        data-key="2"
        disabled={true}
        onKeyDown={[Function]}
        tabIndex={-1}
        value="2"
      >
        Disabled MenuItem
      </div>
    </li>
  </ul>
</div>
`;

exports[`Storyshots DropdownSelector/MenuList Group 1`] = `
.c4 {
  list-style: none;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 40px;
  cursor: pointer;
  outline: none;
  padding-right: 16px;
  padding-left: 16px;
}

.c5:disabled,
.c5[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c5:hover,
.c5:focus,
.c5:focus-within {
  background-color: var(--charcoal-surface3);
}

.c0 {
  padding: 0;
  margin: 0;
}

.c2 {
  display: block;
  color: var(--charcoal-text3);
  font-size: 12px;
  font-weight: bold;
  padding: 12px 0 8px 16px;
}

.c3 {
  padding-left: 0;
  margin: 0;
  box-sizing: border-box;
  list-style: none;
  overflow: hidden;
}

.c1 {
  display: block;
}

<div
  data-dark={false}
>
  <ul
    className="c0"
  >
    <li
      className="c1"
      role="presentation"
    >
      <span
        className="c2"
      >
        Section1
      </span>
      <ul
        className="c3"
        role="group"
      >
        <li
          className="c4"
          role="option"
        >
          <div
            className="c5"
            data-key="0"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={-1}
            value="0"
          >
            Menu 
            0
          </div>
        </li>
        <li
          className="c4"
          role="option"
        >
          <div
            className="c5"
            data-key="1"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={-1}
            value="1"
          >
            Menu 
            1
          </div>
        </li>
        <li
          className="c4"
          role="option"
        >
          <div
            className="c5"
            data-key="2"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={-1}
            value="2"
          >
            Menu 
            2
          </div>
        </li>
        <li
          className="c4"
          role="option"
        >
          <div
            className="c5"
            data-key="3"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={-1}
            value="3"
          >
            Menu 
            3
          </div>
        </li>
        <li
          className="c4"
          role="option"
        >
          <div
            className="c5"
            data-key="4"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={-1}
            value="4"
          >
            Menu 
            4
          </div>
        </li>
      </ul>
    </li>
    <li
      className="c1"
      role="presentation"
    >
      <span
        className="c2"
      >
        Section2
      </span>
      <ul
        className="c3"
        role="group"
      >
        <li
          className="c4"
          role="option"
        >
          <div
            className="c5"
            data-key="5"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={-1}
            value="5"
          >
            Menu 
            5
          </div>
        </li>
        <li
          className="c4"
          role="option"
        >
          <div
            className="c5"
            data-key="6"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={-1}
            value="6"
          >
            Menu 
            6
          </div>
        </li>
        <li
          className="c4"
          role="option"
        >
          <div
            className="c5"
            data-key="7"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={-1}
            value="7"
          >
            Menu 
            7
          </div>
        </li>
        <li
          className="c4"
          role="option"
        >
          <div
            className="c5"
            data-key="8"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={-1}
            value="8"
          >
            Menu 
            8
          </div>
        </li>
        <li
          className="c4"
          role="option"
        >
          <div
            className="c5"
            data-key="9"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={-1}
            value="9"
          >
            Menu 
            9
          </div>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
